core/usage_interceptor.go

```diff
@@ -0,0 +1,157 @@
+package core
+
+import (
+	"context"
+	"fmt"
+
+	grpcm "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/textileio/textile/v2/api/billingd/common"
+	"github.com/textileio/textile/v2/buckets"
+	mdb "github.com/textileio/textile/v2/mongodb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type preFunc func(ctx context.Context, method string) (context.Context, error)
+type postFunc func(ctx context.Context, method string) error
+
+func unaryServerInterceptor(pre preFunc, post postFunc) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		newCtx, err := pre(ctx, info.FullMethod)
+		if err != nil {
+			return nil, err
+		}
+		res, err := handler(newCtx, req)
+		if err != nil {
+			return nil, err
+		}
+		if err = post(newCtx, info.FullMethod); err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+}
+
+func streamServerInterceptor(pre preFunc, post postFunc) grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		newCtx, err := pre(stream.Context(), info.FullMethod)
+		if err != nil {
+			return err
+		}
+		wrapped := grpcm.WrapServerStream(stream)
+		wrapped.WrappedContext = newCtx
+		err = handler(srv, wrapped)
+		if err != nil {
+			return err
+		}
+		return post(newCtx, info.FullMethod)
+	}
+}
+
+func (t *Textile) preUsageFunc(ctx context.Context, method string) (context.Context, error) {
+	for _, ignored := range authIgnoredMethods {
+		if method == ignored {
+			return ctx, nil
+		}
+	}
+	for _, ignored := range usageIgnoredMethods {
+		if method == ignored {
+			return ctx, nil
+		}
+	}
+	var account *mdb.Account
+	if org, ok := mdb.OrgFromContext(ctx); ok {
+		account = org
+	} else if dev, ok := mdb.DevFromContext(ctx); ok {
+		account = dev
+	}
+	// @todo: Account for users after User -> Account migration
+	if account == nil || account.CustomerID == "" {
+		return ctx, nil
+	}
+
+	cus, err := t.bc.GetCustomer(ctx, account.CustomerID)
+	if err != nil {
+		return ctx, err
+	}
+	if err := common.StatusCheck(cus.Status); err != nil {
```

Comment:
As mentioned in other places, this will report an error if the customer status is not in good standing.